include/wil/network.h

```diff
@@ -0,0 +1,1328 @@
+    {
+        const auto& lhs{*this};
+
+        if (lhs.family() != rhs.family())
+        {
+            return lhs.family() < rhs.family();
+        }
+
+        // for operator<, we cannot just memcmp the raw sockaddr values - as they are in network-byte order
+        // we have to first convert back to host-byte order to do comparisons
+        // else the user will see odd behavior, like 1.1.1.1 < 0.0.0.0 (which won't make senses)
+        switch (lhs.family())
+        {
+        case AF_INET:
+        {
+            // compare the address first
+            auto comparison{::memcmp(lhs.in_addr(), rhs.in_addr(), sizeof(IN_ADDR))};
+            if (comparison != 0)
+            {
+                return comparison < 0;
+            }
+
+            // then compare the port (host-byte-order)
+            // only resolve the ntohs() once
+            const auto lhs_port{lhs.port()};
+            const auto rhs_port{rhs.port()};
+            if (lhs_port != rhs_port)
+            {
+                return lhs_port < rhs_port;
+            }
+
+            // must be exactly equal, so not less-than
+            return false;
+        }
+
+        case AF_INET6:
+        {
+            // compare the address first
+            auto comparison{::memcmp(lhs.in6_addr(), rhs.in6_addr(), sizeof(IN6_ADDR))};
+            if (comparison != 0)
+            {
+                return comparison < 0;
+            }
+
+            // then compare the port (host-byte-order)
+            // only resolve the ntohs() once
+            const auto lhs_port{lhs.port()};
+            const auto rhs_port{rhs.port()};
+            if (lhs_port != rhs_port)
+            {
+                return lhs_port < rhs_port;
+            }
+
+            // then compare the scope_id of the address
+            const auto lhs_scope_id{lhs.scope_id()};
+            const auto rhs_scope_id{rhs.scope_id()};
+            if (lhs_scope_id != rhs_scope_id)
+            {
+                return lhs_scope_id < rhs_scope_id;
+            }
+
+            // then compare flow_info
+            const auto lhs_flow_info{lhs.flow_info()};
+            const auto rhs_flow_info{rhs.flow_info()};
+            if (lhs_flow_info != rhs_flow_info)
+            {
+                return lhs_flow_info < rhs_flow_info;
+            }
+
+            // must be exactly equal, so not less-than
+            return false;
+        }
+
+        default:
+            // if not AF_INET or AF_INET6, and families don't match
+            // then just raw memcmp the largest field of the union (v6)
+            return ::memcmp(&lhs.m_sockaddr.Ipv6, &rhs.m_sockaddr.Ipv6, sizeof(SOCKADDR_IN6)) < 0;
+        }
+    }
+
+    inline bool socket_address::operator>(const ::wil::network::socket_address& rhs) const WI_NOEXCEPT
+    {
+        if (*this == rhs)
+        {
+            return false;
+        }
+        return !(*this < rhs);

```

Comment:
Is it possible to write something like a `compare` function (that returns negative for less, zero for equal, and positive for greater)? As-is, you're doing one comparison for the price of two here

Reply:
This is a great idea, thanks! Updated.